task/mgmt-gateway/src/mgs_common.rs

```diff
@@ -94,46 +94,55 @@ impl MgsCommon {
             .prep_image_update(UpdateTarget::Alternate)
             .map_err(|err| ResponseError::UpdateFailed(err as u32))?;
 
-        self.update_buf
-            .start(update.stream_id, update.total_size as usize);
+        self.update_buf.start(update.id, update.total_size);
 
         Ok(())
     }
 
-    pub(crate) fn update_prepare_status(
-        &mut self,
-        request: UpdatePrepareStatusRequest,
-    ) -> Result<UpdatePrepareStatusResponse, ResponseError> {
-        self.update_buf
-            .ensure_matching_stream_id(request.stream_id)?;
-
-        // We immediately prepare for update in `update_prepare()`
-        // and have no followup work to do; if this stream ID
-        // matches, we're already prepared.
-        Ok(UpdatePrepareStatusResponse { done: true })
-    }
-
     pub(crate) fn update_chunk(
         &mut self,
         chunk: UpdateChunk,
         data: &[u8],
     ) -> Result<(), ResponseError> {
         self.update_buf
-            .ingest_chunk(
-                chunk.stream_id,
-                &self.update_task,
-                chunk.offset,
-                data,
-            )
+            .ingest_chunk(&chunk.id, &self.update_task, chunk.offset, data)
             .map(|_progress| ())
     }
 
-    pub(crate) fn update_abort(&mut self) -> Result<(), ResponseError> {
-        self.update_task
-            .abort_update()
-            .map_err(|err| ResponseError::UpdateFailed(err as u32))?;
+    pub(crate) fn status(&self) -> UpdateStatus {
+        self.update_buf.status()
+    }
+
+    pub(crate) fn update_abort(
+        &mut self,
+        id: &UpdateId,
+    ) -> Result<(), ResponseError> {
+        // We will allow the abort if either:
+        //
+        // 1. We have an in-progress update that matches `id`
+        // 2. We do not have an in-progress update
+        //
+        // We only want to return an error if we have a _different_ in-progress
+        // update.
+        match self.update_buf.in_progress_update_id() {
+            Some(in_progress_id) if id != in_progress_id => {
+                return Err(ResponseError::UpdateInProgress(
+                    self.update_buf.status(),
+                ));
+            }
+            _ => (),
+        }
+
+        match self.update_task.abort_update() {
+            // Aborting an update that hasn't started yet is fine; either way
+            // our caller is clear to start a new update.
+            Ok(()) | Err(UpdateError::UpdateNotStarted) => (),
+            Err(other) => {
+                return Err(ResponseError::UpdateFailed(other as u32))
+            }
+        }
```

Comment:
I'm not sure what syntax should work since `let_chains` are still unstable (?!). What if I rolled the rest of the method into the successful branch:

```rust
        match self.update_task.abort_update() {
            // Aborting an update that hasn't started yet is fine; either way
            // our caller is clear to start a new update.
            Ok(()) | Err(UpdateError::UpdateNotStarted) => {
                self.update_buf.abort();
                Ok(())
            }
            Err(other) => {
                Err(ResponseError::UpdateFailed(other as u32))
            }
        }
```